OpenCL/m27000-pure.cl

```diff
@@ -0,0 +1,699 @@
+
+    u32 t = BOX (((d00 >>  0) & 0xff), 4, s_skb)
+           | BOX (((d07 >>  0) & 0xff)
+                 |((d00 >>  8) & 0xff), 5, s_skb)
+           | BOX (((d07 >>  8) & 0xff), 6, s_skb)
+           | BOX (((d21 >>  0) & 0xff)
+                 |((d22 >>  0) & 0xff), 7, s_skb);
+
+    Kc[i] = ((t << 16) | (s & 0x0000ffff));
+    Kd[i] = ((s >> 16) | (t & 0xffff0000));
+  }
+}
+
+DECLSPEC void transform_netntlmv1_key (const u32 w0, const u32 w1, u32 *out)
+{
+  u32 t[8];
+
+  t[0] = (w0 >>  0) & 0xff;
+  t[1] = (w0 >>  8) & 0xff;
+  t[2] = (w0 >> 16) & 0xff;
+  t[3] = (w0 >> 24) & 0xff;
+  t[4] = (w1 >>  0) & 0xff;
+  t[5] = (w1 >>  8) & 0xff;
+  t[6] = (w1 >> 16) & 0xff;
+  t[7] = (w1 >> 24) & 0xff;
+
+  u32 k[8];
+
+  k[0] =               (t[0] >> 0);
+  k[1] = (t[0] << 7) | (t[1] >> 1);
+  k[2] = (t[1] << 6) | (t[2] >> 2);
+  k[3] = (t[2] << 5) | (t[3] >> 3);
+  k[4] = (t[3] << 4) | (t[4] >> 4);
+  k[5] = (t[4] << 3) | (t[5] >> 5);
+  k[6] = (t[5] << 2) | (t[6] >> 6);
+  k[7] = (t[6] << 1);
+
+  out[0] = ((k[0] & 0xff) <<  0)
+         | ((k[1] & 0xff) <<  8)
+         | ((k[2] & 0xff) << 16)
+         | ((k[3] & 0xff) << 24);
+
+  out[1] = ((k[4] & 0xff) <<  0)
+         | ((k[5] & 0xff) <<  8)
+         | ((k[6] & 0xff) << 16)
+         | ((k[7] & 0xff) << 24);
+}
+
+#ifdef KERNEL_STATIC
+DECLSPEC u8 hex_convert (const u8 c)
+{
+  return (c & 15) + (c >> 6) * 9;
+}
+
+DECLSPEC u8 hex_to_u8 (const u8 *hex)
+{
+  u8 v = 0;
+
+  v |= ((u8) hex_convert (hex[1]) << 0);
+  v |= ((u8) hex_convert (hex[0]) << 4);
+
+  return (v);
+}
+#endif
+
+typedef struct netntlm
+{
+  u32 user_len;
+  u32 domain_len;
+  u32 srvchall_len;
+  u32 clichall_len;
+
+  u32 userdomain_buf[64];
+  u32 chall_buf[256];
+
+} netntlm_t;
+
+typedef struct netntlm_tmp
+{
+  u32 digest_buf[4];
+
+} netntlm_tmp_t;
+
+KERNEL_FQ void m27000_init (KERN_ATTR_TMPS_ESALT (netntlm_tmp_t, netntlm_t))
+{
+  /**
+   * modifier
+   */
+
+  const u64 gid = get_global_id (0);
+  const u64 lid = get_local_id (0);
+  const u64 lsz = get_local_size (0);
+
+  if (gid >= gid_max) return;
+
+  /**
+   * salt
+   */
+
+  u32 in[16];
+
+  in[ 0] = pws[gid].i[ 0];
+  in[ 1] = pws[gid].i[ 1];
+  in[ 2] = pws[gid].i[ 2];
+  in[ 3] = pws[gid].i[ 3];
+  in[ 4] = pws[gid].i[ 4];
+  in[ 5] = pws[gid].i[ 5];
+  in[ 6] = pws[gid].i[ 6];
+  in[ 7] = pws[gid].i[ 7];
+
+  u8 *in_ptr = (u8 *) in;
+
+  u32 out[4];
+
+  u8 *out_ptr = (u8 *) out;
+
+  for (int i = 0, j = 0; i < 16; i += 1, j += 2)
+  {
+    out_ptr[i] = hex_to_u8 (in_ptr + j);
+  }
+
+  tmps[gid].digest_buf[0] = out[ 0];
+  tmps[gid].digest_buf[1] = out[ 1];
+  tmps[gid].digest_buf[2] = out[ 2];
+  tmps[gid].digest_buf[3] = out[ 3];
+
+}
+
+KERNEL_FQ void m27000_loop (KERN_ATTR_TMPS_ESALT (netntlm_tmp_t, netntlm_t))
+{
+
+}
+
+KERNEL_FQ void m27000_comp (KERN_ATTR_TMPS_ESALT (netntlm_tmp_t, netntlm_t))
+{
+   /**
+   * modifier
+   */
+
+  const u64 gid = get_global_id (0);
+
+  if (gid >= gid_max) return;
```

Comment:
Ugly race condition here. When working with shared memory we can't return unless all LSZ threads ran through the initializer for() loop otherwise we do not fill all buffers correctly and end up in false negatives. 

As an example, please check m01500_sxx() in OpenCL/m01500_a0-pure.cl

Reply:
removed the return. Have it taking place after the sync.